analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs

```diff
@@ -205,20 +249,23 @@ private ProgramState ProcessInvocation(SymbolicContext context, IInvocationOpera
                     nonEmptyAccess.Add(context.Operation.Instance);
                 }
             }
-            if (instance.Type.DerivesOrImplementsAny(TrackedCollectionTypes))
-            {
-                return ProcessAddMethod(context.State, targetMethod, instance)
-                    ?? ProcessRemoveMethod(context.State, targetMethod, instance)
-                    ?? ProcessClearMethod(context.State, targetMethod, instance)
-                    ?? context.State;
-            }
+            return ProcessAddMethod(context.State, targetMethod, instance)
+                ?? ProcessRemoveMethod(context.State, targetMethod, instance)
+                ?? ProcessClearMethod(context.State, targetMethod, instance)
+                ?? context.State;
         }
         return context.State;
 
         bool HasFilteringPredicate() =>
             invocation.Arguments.Any(x => x.ToArgument().Parameter.Type.Is(KnownType.System_Func_T_TResult));
     }
 
+    private static bool IsWithinTrackedInvocation(IOperationWrapperSonar operation) =>
+        operation.Parent.AsInvocation() is { } invocation
+        && (RaisingMethods.Contains(invocation.TargetMethod.Name)
+            || AddMethods.Contains(invocation.TargetMethod.Name)
+            || RemoveMethods.Contains(invocation.TargetMethod.Name));
```

Comment:
Why this? If it's add or remove method, we should throw away the constraint.
```suggestion
```

Reply:
The logic for that is already in place inside ProcessInvocation. Here we just need to avoid removing the CollectionConstraint if the argument parent invocation was already processed through the ProcessInvocation method before